bin/setupReplication.js

```diff
@@ -0,0 +1,431 @@
+                    errorStack: err.stack,
+                });
+                return cb(err);
+            }
+            this._log.debug('Attached resource policy', {
+                bucket: where,
+                response: res,
+                method: '_attachResourcePolicy',
+            });
+            return cb(null, res);
+        });
+    }
+
+    _enableReplication(roleArns, cb) {
+        const params = {
+            Bucket: this._sourceBucket,
+            ReplicationConfiguration: {
+                Role: roleArns,
+                Rules: [{
+                    Destination: {
+                        Bucket: `arn:aws:s3:::${this._targetBucket}`,
+                    },
+                    Prefix: '',
+                    Status: 'Enabled',
+                }],
+            },
+        };
+        this._s3Clients.source.putBucketReplication(params, (err, res) => {
+            if (err) {
+                this._log.error('error enabling replication', {
+                    error: err,
+                    method: '_SetupReplication._enableReplication',
+                    errorMessage: err.message,
+                    errorStack: err.stack,
+                });
+                return cb(err);
+            }
+            this._log.debug('Bucket replication enabled', {
+                response: res,
+                method: '_enableReplication',
+            });
+            return cb(null, res);
+        });
+    }
+
+    _parallelTasks(cb) {
+        return async.parallel({
+            sourceBucket: next => this._createBucket('source', next),
+            targetBucket: next => this._createBucket('target', next),
+            sourceRole: next => this._createRole('source', next),
+            targetRole: next => this._createRole('target', next),
+            sourcePolicy: next => this._createPolicy('source', next),
+            targetPolicy: next => this._createPolicy('target', next),
+        }, cb);
+    }
+
+    _seriesTasks(data, cb) {
+        const sourceRole = data.sourceRole.Role;
+        const targetRole = data.targetRole.Role;
+        const sourcePolicyArn = data.sourcePolicy.Policy.Arn;
+        const targetPolicyArn = data.targetPolicy.Policy.Arn;
+        const roleArns = `${sourceRole.Arn},${targetRole.Arn}`;
+
+        return async.series([
+            next => this._enableVersioning('source', next),
+            next => this._enableVersioning('target', next),
+            next => this._attachResourcePolicy(sourcePolicyArn,
+                sourceRole.RoleName, 'source', next),
+            next => this._attachResourcePolicy(targetPolicyArn,
+                targetRole.RoleName, 'target', next),
+            next => this._enableReplication(roleArns, next),
+        ], cb);
+    }
+
+    run(cb) {
+        return async.waterfall([
+            next => this._parallelTasks(next),
+            (setupInfo, next) => this._seriesTasks(setupInfo, next),
+            (args, next) => this._checkSanity(next),
```

Comment:
I think the `_checkSanity` function is very useful, though I would think it fits better in an alternative script that checks an existing configuration than here, where the configuration has just been applied, and is supposedly correct because that's the script's job and it already checks for errors (if not correct, it would mean the script is buggy, not that the administrator has failed to configure replication correctly).

We could if we want to run the sanity check script just after this one, but I don't think it's really necessary. @rahulreddy may have another opinion on this.